### PR TITLE
SQR allotment processing & SQR submissions processing

### PR DESCRIPTION
**Pull request by ahmad-elbatanouni**
**1. SQR allotment processing**
1. sends a notification email to the selected devotee using sendInBlue **Node.js** API.
2. sets the file status to **`Given`**.
3. sets the **allotments** sub-object under the allotted file path.

**2. SQR submissions processing**
1. Copies the form submitted data under **`/sqr/submissions`**
2. Updates the allotment according to the rules specified.
3. sends a notification email to the coordinator with all the data of the **submission**, **file**, **devotee's allotments**, a Boolean indicating if this is the first submission by this devotee or not.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -4,7 +4,12 @@ The main credential are automatically set up for you, howver, the **databaseURL*
 Use the following command to set the **databaseURL** & the **storageBucket** URLs:
   
 ```sh
-$ firebase functions:config:set audioseva.database_url="THE DATABASE URL" audioseva.storage_bucket="THE STORAGE BUCKET URL"
+$ firebase functions:config:set audioseva.database_url="THE DATABASE URL"
+$ firebase functions:config:set audioseva.storage_bucket="THE STORAGE BUCKET URL"
+$ firebase functions:config:set audioseva.send_in_blue.key="sendInBlue secret Key"
+# Coordinator details
+$ firebase functions:config:set audioseva.coordinator.templateid=NUMBER
```
Comment (anantakrishna) [inline README.md]: Will not work out. There will be many templates for various notifications.
```suggestion
$ firebase functions:config:set audioseva.sqr.allotment.emailTemplateid=NUMBER
```

File: README.md [hunk 2]
```diff
@@ -4,7 +4,12 @@ The main credential are automatically set up for you, howver, the **databaseURL*
...
+$ firebase functions:config:set audioseva.coordinator.email='EMAIL'
```
Comment (anantakrishna) [inline README.md]:
```suggestion
$ firebase functions:config:set audioseva.coordinator.emailAddress='EMAIL'
```

File: functions/helpers.js [hunk 3]
```diff
@@ -1,38 +1,37 @@
...
+
+    sendEmail = {
+        to: [{ email: to }], // add a name prop if it fails to send
+        bcc,
+        templateId,
+        params
+    };
+
+    apiInstance.sendTransacEmail(sendEmail).then(function(data) {
+        console.log('API called successfully. Returned data: ');
```
Comment (anantakrishna) [inline functions/helpers.js]: Will be too ambiguous which API was called. Be more specific.

File: README.md [hunk 4]
```diff
@@ -4,7 +4,12 @@ The main credential are automatically set up for you, howver, the **databaseURL*
 Use the following command to set the **databaseURL** & the **storageBucket** URLs:
   
 ```sh
-$ firebase functions:config:set audioseva.database_url="THE DATABASE URL" audioseva.storage_bucket="THE STORAGE BUCKET URL"
+$ firebase functions:config:set audioseva.database_url="THE DATABASE URL"
```
Comment (anantakrishna) [inline README.md]: I'm not sure we should use `audioseva`prefix, as all this is about audioseva. Is it a good practice that you learned in Firebase doc?

File: functions/index.js [hunk 5]
```diff
@@ -1,33 +1,21 @@
...
+    });
+} catch (err) { console.log(err) }
+
+//Loading the Functions from their modules
+const glob = require('glob');
+const functionFiles = glob.sync('./src/*.functions.js', { cwd: __dirname });
+
+functionFiles.forEach((file, index) => {
+    const filename = file.split('/').pop().slice(0, -13); // remove the trailing ".functions.js"
+    exports[filename] = require(file);
```
Comment (anantakrishna) [inline functions/index.js]: Add comment here on what will the exported function names.

File: functions/src/SQR.functions.js [hunk 6]
```diff
@@ -0,0 +1,234 @@
...
+const functions = require('firebase-functions');
+const admin = require("firebase-admin");
+
+const bucket = admin.storage().bucket();
+const db = admin.database();
+const helpers = require('./../helpers');
+
+
+
+const SQR_Helper = require('./SQR_Helper');
```
Comment (anantakrishna) [inline functions/src/SQR.functions.js]: Do we really need a separate helper for each module? Why not moving all the SQR-related code into SQR file? Just asking what is behind your decision.

File: functions/src/SQR.functions.js [hunk 7]
```diff
@@ -0,0 +1,234 @@
...
+    .onUpdate((change, context) => {
+        const old = change.before.val();
+        const _new = change.after.val();        
+        let coordinatorConfig = functions.config().audioseva.coordinator;
+        
+        db.ref('/sqr/allotments').once('value')
+        .then(snapshot => {
+            const allotments = snapshot.val();
+            let alloted_num = 0, alloted_before = false;
+            for(let key in allotments) {
```
Comment (anantakrishna) [inline functions/src/SQR.functions.js]: There should be space after `for`. The same for `if` etc. Make consistent everywhere please.

File: functions/src/SQR.functions.js [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+                let allotment = allotments[key];
+                if(allotment.devotee)
+                    if(allotment.devotee.emailAddress)
+                        if(allotment.devotee.emailAddress === _new.devotee.emailAddress)
+                            alloted_before++;
+            }
+            if (alloted_before > 1)
+                alloted_before = true;
+            SQR_Helper.sendEmailOnFileAllotment(coordinatorConfig, old, _new, change.after, alloted_before, helpers.sendEmail)
```
Comment (anantakrishna) [inline functions/src/SQR.functions.js]: Particularly here, you will not send the same allotment email from other places, only from this one place. What is the use of extracting the code into helper?

File: functions/src/SQR.functions.js (hunk 7 above)
Comment (anantakrishna) [reply functions/src/SQR.functions.js]: And more importantly, are you really iterating all the allotments? Consider querying in order to get previous allotments and submissions.

File: functions/src/SQR.functions.js [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+        db.ref(`/sqr/submissions/${original.serial}`).update(submission);
+
+
+
+        // 2. Update the allotment
+        let allotmentUpdates = { status: audioFileStatus };
+
+        // in case 1 & 2 add the comments to the notes
+        if (audioFileStatus !== 'WIP')
+            allotmentUpdates.notes = original.comments;
```
Comment (anantakrishna) [inline functions/src/SQR.functions.js]: Don't overwrite the notes, append instead.

File: functions/src/SQR.functions.js [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+
+
+
+        // Coordinator object example { templateid: 3, email:'a@a.a', name: 'Aj' }
+        let coordinator = functions.config().audioseva.coordinator;
+        // 3. Notify the coordinator
+        // 3.1 Get the submitted audio file data
+
+        //  EXTRACTING the list name first from the file_name
+        let list = original.audio_file_name.match(/^[^-]*[^ -]/g)[0];
```
Comment (anantakrishna) [inline functions/src/SQR.functions.js]: This is a good candidate for a helper, as you will be extracting list many times from various places.

File: functions/src/SQR.functions.js [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        let file_name = original.audio_file_name.slice( list.length + 1, original.audio_file_name.length );
```
Comment (anantakrishna) [inline functions/src/SQR.functions.js]: No need. Filename should include the list.. Don't cut it.

File: functions/src/SQR.functions.js [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+        
+        db.ref(`/sqr/files/${list}/${file_name}`).once('value')
+        .then(snapshot => {
+            if(snapshot.exists()) {
+                let fileData = snapshot.val();
+
+                // 3.2 Get the devotee's Allotments in ('given' || 'WIP') state
+                // Using the /sqr/files here instead of /sqr/allotmens to be able to check for the 
+                // state of ('given' || 'WIP')
+                db.ref(`/sqr/files`).once('value')
```
Comment (anantakrishna) [inline functions/src/SQR.functions.js]: Again, iterating files might be not a good way. Consider querying.

File: functions/src/SQR_Helper.js [hunk 13]
```diff
@@ -0,0 +1,148 @@
...
+
+    return 1;    
+}
+
+exports.sendEmailOnFileAllotment = function  (coordinatorConfig, old, _new, new_snapshot, alloted_before, sendEmail) {
+
+    if(!old.filesAlloted && _new.filesAlloted)
+        if (_new.devotee)
+            if(_new.devotee.emailAddress) {
+                let email = _new.devotee.emailAddress;
```
Comment (anantakrishna) [inline functions/src/SQR_Helper.js]: Are you sure we need these variables that are used only once?

File: README.md (hunk 1 above)
Comment (ahmad-elbatanouni) [reply README.md]: wouldn't it be better if I specify the kind of email recipient ( devotee or coordinator ) ?
**`sqr.allotment.emailTemplateid`** can be somehow ambiguous

File: README.md (hunk 2 above)
Comment (ahmad-elbatanouni) [reply README.md]: Config data can NOT have upper-cased names, can I replace it with something like this  **`coordinator.email_address`**

File: functions/helpers.js (hunk 3 above)
Comment (ahmad-elbatanouni) [reply functions/helpers.js]: This was just for testing purposes, but I'll log everything just in case it will be needed later.

File: README.md (hunk 4 above)
Comment (ahmad-elbatanouni) [reply README.md]: No not really, in fact they suggested prepending the key with the service name ( this only applies to the sendInBlue Key ) .. So I think it really needs restructuring.

File: functions/index.js (hunk 5 above)
Comment (ahmad-elbatanouni) [reply functions/index.js]: I'm sorry I didn't understand what you mean here

File: functions/src/SQR.functions.js (hunk 6 above)
Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: No, you're right ... in the beginning I thought it would be better to keep the SQR functions clean by having the calling separate functions with names describing what they do .. like this for example
```javascript
exports.syncStorageToDB = functions.https.onRequest((req, res) => {
    SQR_Helper.handleCurrentlyUploadedFiles(bucket, db, helpers.storeFileNameToDB);
    SQR_Helper.removeNonExistingMp3DBEntries(bucket, db, helpers.removeFileNameFromDB);
    return res.send(`Started Execution, the process is now Running in the background`);
});
```

but I think you're right .. it's useless, I can replace this with comments before every function

File: functions/src/SQR.functions.js (hunk 12 above)
Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: In this one particularly, I'm unable to do it.. the structure is as follows:

**--files
----lists [array]
--------file
----------allotment
------------devotee
--------------emailAddress**

what I want to look for is the **emailAddress**, 
In the security rules I added a new index
```
"sqr": {
      "files": {
        "$list": {
         	".indexOn": ["status", "allotment/devotee/emailAddress"]
        }
      },
}
```
and in the JS functions, I used 
```javascript
db.ref(`/sqr/files`).orderByChild('allotment/devotee/emailAddress').equalTo('a@a.a')
```
The problem is that **list name** is variable, is there a way to automatically let Firebase know that (like in the security rules where **$list** is used)?

File: functions/index.js (hunk 5 above)
Comment (anantakrishna) [reply functions/index.js]: Add a block comment before this code and describe how the function name is constructed from the  file name and the JS function. You can give an example.

File: functions/src/SQR.functions.js (hunk 6 above)
Comment (anantakrishna) [reply functions/src/SQR.functions.js]: Yes. And consider using classes if possible and reasonable.

File: functions/src/SQR.functions.js (hunk 7 above)
Comment (anantakrishna) [reply functions/src/SQR.functions.js]: AFAIK it's possible. However consider creating an index in the db rules file in the root.

File: functions/src/SQR.functions.js (hunk 12 above)
Comment (anantakrishna) [reply functions/src/SQR.functions.js]: I cannot understand the problem. Can you rephrase the question?

Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: I want to search for the **emailAddress** ... this is located in the following structure:

--files
----lists [array]
--------**file**
----------allotment
------------devotee
--------------emailAddress
if the **`file`** ( above in the data strcture ) had no **`lists`** parent I'd have searched like that
```javascript
db.ref(`/sqr/files`).orderByChild('allotment/devotee/emailAddress').equalTo('a@a.a')
```
of course after adding **`".indexOn":  ["allotment/devotee/emailAddress"]`** at **`/sqr/files`** to the security rules.

the problem here is the existence of the **lists** parent, I don't want to make the index on the 
**`/sqr/files`** .. I wan't it to be on **`/sqr/files/list_name`**
this can be achieved in the security rules by having something like this 
```
"sqr": {
      "files": {
        "$list": {
         	".indexOn": ["allotment/devotee/emailAddress"]
        }
      },
}
```

what I couldn't find online is how can I use this new index in the node.js API .. how can I add something like **`$list`** to the code to automatically skip directly to the files and search them on this index **`allotment/devotee/emailAddress`**
```javascript
db.ref(`/sqr/files`).orderByChild('allotment/devotee/emailAddress').equalTo('a@a.a')
```

File: functions/src/SQR.functions.js (hunk 6 above)
Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: in Node.js 8 ( the latest in Firebase functions ) I can't use the **`export`** or **`import`** keywords... So I don't think it would make sense to use classes now.

**Comment by ahmad-elbatanouni** [comment]
I made the requested changes .. only [this one ](https://github.com/PureBhaktiArchive/audioseva/pull/9#discussion_r230364833) remains

File: functions/src/SQR.functions.js (hunk 12 above)
Comment (anantakrishna) [reply functions/src/SQR.functions.js]: Why don't you want to add index on the files node?

Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: Tried that already but didn't work out... it returns **`null`**
**security rules:**
g1


**db entry:**
g2

**JS:**
g3

**result:**
g4

Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: Tried that already, but it returns **`null`**

**1. Security Rules**
g1

**2. DB entry**
g2


**3. JS**
g3


**4. Result**
g4

File: functions/src/SQR.functions.js [hunk 14]
```diff
@@ -20,33 +18,64 @@ const SQR_Helper = require('./SQR_Helper');
...
+            }).catch(err => console.log(err));
+        });
+
+        return 1;
 });
 
 exports.sendEmailOnFileAllotment = functions.database.ref('/sqr/allotments/{allotment_id}')
     .onUpdate((change, context) => {
         const old = change.before.val();
         const _new = change.after.val();
```
Comment (anantakrishna) [inline functions/src/SQR.functions.js]: I don't like the underscore here. And something more obvious would be better, like `newAllotment` or at least `newData`.

File: functions/src/SQR.functions.js [hunk 15]
```diff
@@ -83,7 +142,20 @@ exports.syncStorageToDB = functions.https.onRequest((req, res) => {
 
 
 exports.importMP3IntoSQR = functions.storage.object().onFinalize( object => {
```
Comment (anantakrishna) [inline functions/src/SQR.functions.js]: Better to name it `importFilesFromStorage`.

File: functions/src/SQR.functions.js (hunk 12 above)
Comment (anantakrishna) [reply functions/src/SQR.functions.js]: Yes, I see that it is not possible to use wildcards in the `orderBy` function.
We'll sort it out later, please remove this code for now and just pass empty array to the email template.

**Comment by anantakrishna** [comment]
In the future please don't mix tasks in one branch and pull request.

File: functions/src/SQR.functions.js (hunk 12 above)
Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: This was just an example for explaining my point ... of course in the original code I add the **email of the devotee**.
What that was originally about is me **iterating** the allotments in all the files, you asked me then to do it by querying ... so I just wanted to show that querying is not possible AFAIK in this partiucular case.

File: README.md [hunk 16]
```diff
@@ -4,12 +4,13 @@ The main credential are automatically set up for you, howver, the **databaseURL*
...
+# each arg must have a 2-part key (e.g foo.bar)
+$ firebase functions:config:set sqr.database_url="THE DATABASE URL"
+$ firebase functions:config:set sqr.storage_bucket="THE STORAGE BUCKET URL"
+$ firebase functions:config:set send_in_blue.key="sendInBlue secret Key"
 # Coordinator details
-$ firebase functions:config:set audioseva.coordinator.templateid=NUMBER
-$ firebase functions:config:set audioseva.coordinator.email='EMAIL'
-$ firebase functions:config:set audioseva.coordinator.name='NAME'
+$ firebase functions:config:set sqr.allotment.templateid=NUMBER
+$ firebase functions:config:set sqr.coordinator.email_address='EMAIL'
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply README.md]: Ok

File: README.md [hunk 17]
```diff
@@ -4,12 +4,13 @@ The main credential are automatically set up for you, howver, the **databaseURL*
...
+$ firebase functions:config:set sqr.coordinator.name='NAME'
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply README.md]: Ok

File: functions/index.js [hunk 18]
```diff
@@ -5,17 +5,31 @@ const admin = require("firebase-admin");
 try {
     admin.initializeApp({
         credential: admin.credential.applicationDefault(),
-        databaseURL: functions.config().audioseva.database_url,
-        storageBucket: functions.config().audioseva.storage_bucket
+        databaseURL: functions.config().sqr.database_url,
+        storageBucket: functions.config().sqr.storage_bucket
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply functions/index.js]: Replaced that entirely with an environment variable **`process.env.FIREBASE_CONFIG`**

File: functions/index.js [hunk 19]
```diff
@@ -5,17 +5,31 @@ const admin = require("firebase-admin");
...
+**
+**********************************************/
 const glob = require('glob');
 const functionFiles = glob.sync('./src/*.functions.js', { cwd: __dirname });
 
 functionFiles.forEach((file, index) => {
     const filename = file.split('/').pop().slice(0, -13); // remove the trailing ".functions.js"
-    exports[filename] = require(file);
+    console.log(require(file));
+    // exports[filename] = require(file);
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply functions/index.js]: sorry, forgot to clean the file after debugging.

File: functions/src/SQR.functions.js [hunk 20]
```diff
@@ -20,33 +18,64 @@ const SQR_Helper = require('./SQR_Helper');
...
+                            }
+                        }, err => {
+                            if (!err)
+                                db.ref(`/sqr/allotments/${newDocKey}`).update({ filesAlloted: true });
+                        });
+                return 1;
+            }).catch(err => console.log(err));
+        });
+
+        return 1;
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: the function crashes when nothing is returned. I can either return a promise or a value .. in some functions there are multiple operations ... I didn't know which one to choose to return.

File: functions/src/SQR.functions.js [hunk 21]
```diff
@@ -20,33 +18,64 @@ const SQR_Helper = require('./SQR_Helper');
...
 });
 
 exports.sendEmailOnFileAllotment = functions.database.ref('/sqr/allotments/{allotment_id}')
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: ok

File: functions/src/SQR.functions.js (hunk 14 above)
Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: ok

File: functions/src/SQR.functions.js (hunk 15 above)
Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: ok

File: functions/src/SQR.functions.js [hunk 22]
```diff
@@ -20,33 +18,64 @@ const SQR_Helper = require('./SQR_Helper');
...
+                        let date = new Date();
+                        helpers.sendEmail(
+                            _new.devotee.emailAddress, //email
+                            [{ email: coordinatorConfig.email, name: coordinatorConfig.name }], //bcc
+                            templateId,
+                            { //parameters
+                                files: _new.files,
+                                devotee: _new.devotee,
+                                comment: _new.comment,
+                                date: `${date.getUTCDate() + 1}.${date.getUTCMonth() + 1}`,
```
[the remark this answers is not in the document]
Comment (ahmad-elbatanouni) [reply functions/src/SQR.functions.js]: ok

File: functions/src/SQR.functions.js (hunk 20 above)
[the remark this answers is not in the document]
Comment (anantakrishna) [reply functions/src/SQR.functions.js]: Ok, let's wait for a TypeScript, it should be easier there.

**Comment by anantakrishna** [comment]
[SQR: Allotment processing](https://trello.com/c/hEX4y4zd/12-sqr-allotment-processing)
